src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+ * `gpii.diff.compareObjects` should be used instead.
+ *
+ * @param leftValue - The left-hand value.
+ * @param rightValue - The right-hand value.
+ * @param isArray - Whether we are dealing with arrays or not
+ * @return {Array} - An array of segments representing everything that has changed (or not changed).
+ *
+ */
+gpii.diff.singleValueDiff = function (leftValue, rightValue, isArray) {
+    var key = isArray ? "arrayValue" : "value";
+    var segments = [];
+    if ((isArray && gpii.diff.arraysEqual(leftValue, rightValue)) || (!isArray && leftValue === rightValue)) {
+        var unchangedSegment = { type: "unchanged"};
+        unchangedSegment[key] = leftValue;
+        segments.push(unchangedSegment);
+    }
+    else {
+        if ((!isArray && leftValue !== undefined) || (isArray && leftValue.length)) {
+            var removedSegment = { type: "removed" };
+            removedSegment[key] = leftValue;
+            segments.push(removedSegment);
+        }
+        if ((!isArray && rightValue !== undefined) || (isArray && rightValue.length)) {
+            var addedSegment = { type: "added"};
+            addedSegment[key] = rightValue;
+            segments.push(addedSegment);
+        }
+    }
+    return segments;
+};
+
+/**
+ *
+ * Find the longest matching "phrase" in two strings, which prefers one or more whole words.
+ *
+ * @param leftString {String} - One of the two strings, which will be used to search for matching patterns.
+ * @param rightString {String} - The other string, which will be compared against all possible sub-patterns.
+ * @return {Object} - See below.
+ *
+ * The object returned has three keys:
+ *
+ * `leftIndex`:  The position of the matching segment in `leftString`.  If there is no matching segment, this will be `-1`.
+ * `rightIndex`: The position of the matching segment in `rightString`.  If there is no matching segment, this will be `-1`.
+ * `segment`:    The longest matching string common to both strings.  If there are multiple matching strings of the same length, the first is returned.
+ *
+ */
+gpii.diff.longestCommonStringPhrase = function (leftString, rightString) {
+    var matchDef = {
+        leftIndex: -1,
+        rightIndex: -1,
+        segment: ""
+    };
+
+    if (leftString && rightString) {
+        var leftWordPhraseDefs = gpii.diff.extractPhraseSegments(leftString);
+
+        var firstMatchDef = fluid.find(leftWordPhraseDefs, function (phraseDef) {
+            var rightIndex = rightString.indexOf(phraseDef.value);
+            if (rightIndex !== -1) {
+                return {
+                    segment: phraseDef.value,
+                    leftIndex: phraseDef.index,
+                    rightIndex: rightIndex
+                };
+            }
+        });
+        if (firstMatchDef) {
+            matchDef = firstMatchDef;
+        }
+    }
+    return matchDef;
+};
+
+gpii.diff.sortPhraseDefsByLength = function (a, b) {
```

Comment:
This comparator does more than is advertised in the comments to "extractPhraseSegments" (it breaks ties based on index position) - this should be noted and explained

Reply:
I have renamed it and added JSDocs.